app/src/main/java/com/lagradost/cloudstream3/plugins/PluginManager.kt

```diff
@@ -265,16 +267,98 @@ object PluginManager {
         }
 
         main {
-            createNotification(activity, updatedPlugins)
+            val uitext = txt(R.string.plugins_updated, updatedPlugins.size)
+            createNotification(activity, uitext, updatedPlugins)
         }
 
-       // ioSafe {
+        // ioSafe {
             afterPluginsLoadedEvent.invoke(true)
-       // }
+        // }
 
         Log.i(TAG, "Plugin update done!")
     }
 
+    /**
+     * Automatically download plugins not yet existing on local
+     * 1. Gets all online data from online plugins repo
+     * 2. Fetch all not downloaded plugins
+     * 3. Download them and reload plugins
+     **/
+    fun downloadNotExistingPluginsAndLoad(activity: Activity) {
+        val newDownloadPlugins = mutableListOf<String>()
+        val urls = (getKey<Array<RepositoryData>>(REPOSITORIES_KEY)
+            ?: emptyArray()) + PREBUILT_REPOSITORIES
+        val onlinePlugins = urls.toList().apmap {
+            getRepoPlugins(it.url)?.toList() ?: emptyList()
+        }.flatten().distinctBy { it.second.url }
+
+        val providerLang = activity.getApiProviderLangSettings()
+        //Log.i(TAG, "providerLang => ${providerLang.toJson()}")
+
+        // Iterate online repos and returns not downloaded plugins
+        val notDownloadedPlugins = onlinePlugins.mapNotNull { onlineData ->
+            val sitePlugin = onlineData.second
+            //Don't include empty urls
+            if (sitePlugin.url.isBlank()) { return@mapNotNull null }
+            if (sitePlugin.repositoryUrl.isNullOrBlank()) { return@mapNotNull null }
+
+            //Omit already existing plugins
+            if (getPluginPath(activity, sitePlugin.internalName, onlineData.first).exists()) {
+                Log.i(TAG, "Skip > ${sitePlugin.internalName}")
+                return@mapNotNull null
+            }
+
+            //Omit lang not selected on language setting
+            val lang = sitePlugin.language ?: return@mapNotNull null
+            //If set to 'universal', don't skip any language
+            if (!providerLang.contains(AllLanguagesName) && !providerLang.contains(lang)) {
+                return@mapNotNull null
+            }
+            //Log.i(TAG, "sitePlugin lang => $lang")
+
+            //Omit NSFW, if disabled
+            sitePlugin.tvTypes?.let { tvtypes ->
+                if (!settingsForProvider.enableAdult) {
+                    if (tvtypes.contains(TvType.NSFW.name)) {
```

Comment:
pretty much everything is great besides unnecessary nested if statements here 
`if (!settingsForProvider.enableAdult && tvtypes.contains(TvType.NSFW.name)) {`
does the same :pray:

Reply:
Micro optimization. If the toggle is off, no need to check if tvtype contains the NSFW type.

Reply:
No, the compiler already does that